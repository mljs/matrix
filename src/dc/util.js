export function hypotenuse(a, b) {
  var r = 0;
  if (Math.abs(a) > Math.abs(b)) {
    r = b / a;
    return Math.abs(a) * Math.sqrt(1 + r * r);
  }
  if (b !== 0) {
    r = a / b;
    return Math.abs(b) * Math.sqrt(1 + r * r);
  }
  return 0;
}

export function getFilled2DArray(rows, columns, value) {
  var array = new Array(rows);
  for (var i = 0; i < rows; i++) {
    array[i] = new Array(columns);
    for (var j = 0; j < columns; j++) {
      array[i][j] = value;
    }
  }
  return array;
}
