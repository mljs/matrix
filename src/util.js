/**
 * @private
 * Check that a row index is not out of bounds
 * @param {Matrix} matrix
 * @param {number} index
 * @param {boolean} [outer]
 */
export function checkRowIndex(matrix, index, outer) {
  let max = outer ? matrix.rows : matrix.rows - 1;
  if (index < 0 || index > max) {
    throw new RangeError('Row index out of range');
  }
}

/**
 * @private
 * Check that a column index is not out of bounds
 * @param {Matrix} matrix
 * @param {number} index
 * @param {boolean} [outer]
 */
export function checkColumnIndex(matrix, index, outer) {
  let max = outer ? matrix.columns : matrix.columns - 1;
  if (index < 0 || index > max) {
    throw new RangeError('Column index out of range');
  }
}

/**
 * @private
 * Check that the provided vector is an array with the right length
 * @param {Matrix} matrix
 * @param {Array|Matrix} vector
 * @return {Array}
 * @throws {RangeError}
 */
export function checkRowVector(matrix, vector) {
  if (vector.to1DArray) {
    vector = vector.to1DArray();
  }
  if (vector.length !== matrix.columns) {
    throw new RangeError(
      'vector size must be the same as the number of columns',
    );
  }
  return vector;
}

/**
 * @private
 * Check that the provided vector is an array with the right length
 * @param {Matrix} matrix
 * @param {Array|Matrix} vector
 * @return {Array}
 * @throws {RangeError}
 */
export function checkColumnVector(matrix, vector) {
  if (vector.to1DArray) {
    vector = vector.to1DArray();
  }
  if (vector.length !== matrix.rows) {
    throw new RangeError('vector size must be the same as the number of rows');
  }
  return vector;
}

export function checkIndices(matrix, rowIndices, columnIndices) {
  return {
    row: checkRowIndices(matrix, rowIndices),
    column: checkColumnIndices(matrix, columnIndices),
  };
}

export function checkRowIndices(matrix, rowIndices) {
  if (typeof rowIndices !== 'object') {
    throw new TypeError('unexpected type for row indices');
  }

  let rowOut = rowIndices.some((r) => {
    return r < 0 || r >= matrix.rows;
  });

  if (rowOut) {
    throw new RangeError('row indices are out of range');
  }

  if (!Array.isArray(rowIndices)) rowIndices = Array.from(rowIndices);

  return rowIndices;
}

export function checkColumnIndices(matrix, columnIndices) {
  if (typeof columnIndices !== 'object') {
    throw new TypeError('unexpected type for column indices');
  }

  let columnOut = columnIndices.some((c) => {
    return c < 0 || c >= matrix.columns;
  });

  if (columnOut) {
    throw new RangeError('column indices are out of range');
  }
  if (!Array.isArray(columnIndices)) columnIndices = Array.from(columnIndices);

  return columnIndices;
}

export function checkRange(matrix, startRow, endRow, startColumn, endColumn) {
  if (arguments.length !== 5) {
    throw new RangeError('expected 4 arguments');
  }
  checkNumber('startRow', startRow);
  checkNumber('endRow', endRow);
  checkNumber('startColumn', startColumn);
  checkNumber('endColumn', endColumn);
  if (
    startRow > endRow ||
    startColumn > endColumn ||
    startRow < 0 ||
    startRow >= matrix.rows ||
    endRow < 0 ||
    endRow >= matrix.rows ||
    startColumn < 0 ||
    startColumn >= matrix.columns ||
    endColumn < 0 ||
    endColumn >= matrix.columns
  ) {
    throw new RangeError('Submatrix indices are out of range');
  }
}

export function newArray(length, value = 0) {
  let array = [];
  for (let i = 0; i < length; i++) {
    array.push(value);
  }
  return array;
}

function checkNumber(name, value) {
  if (typeof value !== 'number') {
    throw new TypeError(`${name} must be a number`);
  }
}

export function checkNonEmpty(matrix) {
  if (matrix.isEmpty()) {
    throw new Error('Empty matrix has no elements to index');
  }
}
