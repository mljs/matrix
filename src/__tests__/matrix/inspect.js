import { inspect } from 'util';

import { Matrix } from '../..';

describe('custom Node.js inspect function', () => {
  it('should work with a simple matrix', () => {
    expect(
      inspect(
        new Matrix([
          [0, 10, 200],
          [-0.3, -0.44, -0.5555],
        ]),
      ),
    ).toMatchSnapshot();
  });
  it('should properly format numbers', () => {
    const a10 = Array.from({ length: 10 }, (_, i) => i);
    expect(
      inspect(
        new Matrix([
          a10,
          a10.map((e) => 0.123456789 * 10 ** e),
          a10.map((e) => 0.123456789 / 10 ** e),
          a10.map((e) => 0.123 * 10 ** e),
          a10.map((e) => 0.123 / 10 ** e),
          a10.map((e) => 0.12 * 10 ** e),
          a10.map((e) => 0.12 / 10 ** e),
        ]),
      ),
    ).toMatchSnapshot();
    expect(
      inspect(
        new Matrix([
          a10,
          a10.map((e) => (-1) ** e * 0.123456789 * 10 ** e),
          a10.map((e) => ((-1) ** e * -0.123456789) / 10 ** e),
          a10.map((e) => (-1) ** e * 0.123 * 10 ** e),
          a10.map((e) => ((-1) ** e * -0.123) / 10 ** e),
          a10.map((e) => (-1) ** e * 0.12 * 10 ** e),
          a10.map((e) => ((-1) ** e * -0.12) / 10 ** e),
        ]),
      ),
    ).toMatchSnapshot();
  });
});
