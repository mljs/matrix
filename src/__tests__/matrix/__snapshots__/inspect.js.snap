// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom Node.js inspect function should properly format numbers 1`] = `
"Matrix {
  [
    0        1        2        3        4        5        6        7        8        9       
    0.123457 1.234568 12.34568 123.4568 1234.568 12345.68 123456.8 1234568  12345679 1.235e+8
    0.123457 0.012346 0.001235 1.235e-4 1.235e-5 1.235e-6 1.235e-7 1.235e-8 1.235e-9 1.23e-10
    0.123    1.23     12.3     123      1230     12300    123000   1230000  12300000 1.230e+8
    0.123    0.0123   0.00123  0.000123 1.230e-5 1.230e-6 1.23e-7  1.23e-8  1.230e-9 1.23e-10
    0.12     1.2      12       120      1200     12000    120000   1200000  12000000 1.200e+8
    0.12     0.012    0.0012   1.200e-4 0.000012 1.200e-6 1.2e-7   1.2e-8   1.2e-9   1.2e-10 
  ]
  rows: 7
  columns: 10
}"
`;

exports[`custom Node.js inspect function should properly format numbers 2`] = `
"Matrix {
  [
     0        1        2        3        4        5        6        7        8        9      
     0.12346 -1.23457  12.3457 -123.457  1234.57 -12345.7  123457  -1234568  1.23e+7 -1.23e+8
    -0.12346  0.01235 -0.00123  1.23e-4 -0.00001  1.23e-6 -0.00000  1.23e-8 -0.00000  1.2e-10
     0.123   -1.23     12.3    -123      1230    -12300    123000  -1230000  1.23e+7 -1.23e+8
    -0.123    0.0123  -0.00123  1.23e-4 -0.00001  1.23e-6 -1.23e-7  1.23e-8 -0.00000  1.2e-10
     0.12    -1.2      12      -120      1200    -12000    120000  -1200000  1.20e+7 -1.20e+8
    -0.12     0.012   -0.0012   1.20e-4 -0.00001  1.20e-6 -1.2e-7   1.2e-8  -1.2e-9   1.2e-10
  ]
  rows: 7
  columns: 10
}"
`;

exports[`custom Node.js inspect function should work with a simple matrix 1`] = `
"Matrix {
  [
     0        10       200    
    -0.3     -0.44    -0.5555 
  ]
  rows: 2
  columns: 3
}"
`;
